### PR TITLE
fix(aio): improve search index to capture better results

### LINKED ISSUE #17545: Searching docs for "validator" yields wrong result
## I'm submitting a ...

<pre><code>
[x ] Documentation issue or request
</code></pre>

## Current behavior

Searching for the term "validator" on https://angular.io/docs (upper right search bar), yields an incomplete result:
 -  [I] ValidatorFn

## Expected behavior

Expecting:

#### under "API":
 -  [I] ValidatorFn
 -  [T] ValidatorErrors
 -  [C] Validators  **<-- probably sought by most**
 -  [I] Validator
 -  [I] AsyncValidator
 -  [D] CheckboxRequiredValidator
 -  ...
 -  [K] NG_ASYNC_VALIDATORS
 -  ... 

#### under "GUIDE" (less important than accurate API results):
 - Form Validation
 - Forms
 - Reactive Forms
 - Dynamic forms
 - ... 

#### under "TUTORIAL" (less important than accurate API results):
 - The Hero Editor

## Minimal reproduction of the problem with instructions

1) visit https://angular.io/api/forms/Validators
2) type "validator" into the upper right search bar
 [note incomplete results]
 [compare with the search term "valid" which gives better results]

## What is the motivation / use case for changing the behavior?

 - The current search result is not what presumably 99% of users are after.
 - Validation related documentation should be easy to find via the search bar.

## Please tell us about your environment

<pre><code>
https://angular.io/ website on 6/15/17

Browser:
- [ X] Chrome (desktop) version 58.0

### PR DESCRIPTION
**Pull request by petebacondarwin**
Fixes #17545 #17889

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Before: 
screen shot 2017-06-16 at 20 26 56

**Comment by petebacondarwin** [comment]
After:

screen shot 2017-06-16 at 20 27 10

File: aio/src/app/search/search-worker.js [hunk 1]
```diff
@@ -86,8 +86,11 @@ function queryIndex(query) {
     terms.forEach(function(term) {
       // Only include terms that are longer than 2 characters, if there is more than one term
       // Add trailing wildcard to each term so that it will match more results
+      // Allow fuzzy matching to improve results
       if (terms.length === 1 || term.trim().length > 2) {
+        qb.term(term, { usePipeline: true });
         qb.term(term, { wildcard: lunr.Query.wildcard.TRAILING });
```
Comment (gkalpak) [inline aio/src/app/search/search-worker.js]: Would  it help fnd more relevant results if we used a leading `*` too?

**Review by gkalpak — COMMENTED** [review]
I feel we are getting too many results now :grin: (but maybe I am wrong). For example the term "Date" gives `Meta` as the first result, because it contains an `updateXyz` method.

I don't know what lunr's capabilities are and if we can get more accurate ranking 😕 

##
Also, I am not sure this fixes the underlying cause of #17545. Why weren't we previously matching the `Validator` interface to the "validator" term (they match 1-1)?

**Comment by petebacondarwin** [comment]
How about we run the previous search and then this more relaxed search if
the results list is small?

**Comment by gkalpak** [comment]
Sounds promising :smiley:

**Comment by petebacondarwin** [comment]
The last improvement needed to get things like the `ngModules` guide to the top of the results is a bug fix in lunr: https://github.com/olivernn/lunr.js/pull/281

But we should merge this PR for now and then update the Lunr dependency when @olivernn lands that fix to Lunr.

**Review by gkalpak — APPROVED** [review]
This is obviously much better than before, but I still find the behavior a little odd. For example as you type `animations`, results are shown and hidden: 

- `anim` -> Many results
- `anima` -> No results
- `animat` -> One result
- `animati` -> No results
- `animatio` -> No results
- `animation` -> Many results

I assume this is not something that olivernn/lunr.js#281 would fix, right?

##
This PR is a very good step forward anyway though. So 👍

**Comment by petebacondarwin** [comment]
> For example as you type animations, results are shown and hidden:
> 
> anim -> Many results
> anima -> No results
> animat -> One result
> animati -> No results
> animatio -> No results
> animation -> Many results

This is not fixed by the changes to lunr.js. It is related to how the stemmer works. The actual queries will look a bit like this:

anim -> anim(ate)
anima -> anima
animat -> animat(e)
animati -> animati
animatio -> animatio
animation -> animat(ion)

I was thinking that a bit of debounce (even 200ms) would improve the UX as the query will only happen after the user finishes typing and so will hide these unfortunate results.

**Comment by olivernn** [comment]
The problem is due to the stemming of the terms, you can _probably_ get better results by constructing the query slightly differently, searching with a wildcard as well as exact matches.

Something like the following would probably work:

```javascript
idx.query(function (q) {
  q.term("anima", { boost: 100 })  // exact match
  q.term("anima", { boost: 10, wildcard: lunr.Query.wildcard.TRAILING, usePipeline: false }) // prefix search, no stemming
  q.term("anima", { editDistance: 1, usePipeline: false }) // fuzzy search, no stemming
})
```

You can try this out on the demo by searching for "anima anima* anima~1". The search with edit distance is optional but will increase recall.

**Comment by petebacondarwin** [comment]
@olivernn I did originally try building my own queries, using a similar mix of wildcards and stemming, etc, but in the end I found that your query parser does a better job than I could.

**Comment by petebacondarwin** [comment]
Just waiting for Travis to approve.

**Comment by petebacondarwin** [comment]
OK, so this PR needs to sit on top of the #17857

**Comment by Foxandxss** [comment]
Guide changes are good for me.

**Review by Foxandxss — APPROVED** [review]

**Comment by petebacondarwin** [comment]
I need to wait for #17857 to land and then rebase before we can land this.

**Comment by petebacondarwin** [comment]
Cherry picked to 4.2.x as 

dbfd7179a8fb3c44ffc64688b12ab72bbc41a797 build(aio): add terms from heading to the search index
207e2febb302b6ed372a6dfe4d2c6245ae9c9d5b build(aio): capture all the headings from a doc in the vFile.headings property
39c9860b9d4fdbca4218de7989b59171956e92da build(aio): revert to general purpose search algorithm

**Pull request merged**